ecs-agent/acs/session/refresh_credentials_responder_test.go

```diff
@@ -0,0 +1,246 @@
+	testRefreshCredentialsMessage.RoleType = tempRoleType
+}
+
+// TestValidateRefreshMessageSuccess tests if a valid credentials message
+// is validated without any errors.
+func TestValidateRefreshMessageSuccess(t *testing.T) {
+	err := validateIAMRoleCredentialsMessage(testRefreshCredentialsMessage)
+	assert.NoError(t, err, "Error validating credentials message: %w", err)
+}
+
+// TestRefreshCredentialsAckHappyPath tests the happy path for a typical IAMRoleCredentialsMessage and confirms expected
+// ACK request is made.
+func TestRefreshCredentialsAckHappyPath(t *testing.T) {
+	testCases := []struct {
+		name     string
+		roleType string
+	}{
+		{
+			name:     "task role type",
+			roleType: credentials.ApplicationRoleType,
+		},
+		{
+			name:     "execution role type",
+			roleType: credentials.ExecutionRoleType,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			// WaitGroup is necessary to wait for function to be called in separate goroutine before exiting the test.
+			wg := sync.WaitGroup{}
+			wg.Add(1)
+
+			iamRoleCredentialsMessageCopy := *testRefreshCredentialsMessage
+			ackSent := make(chan *ecsacs.IAMRoleCredentialsAckRequest)
+			credentialsManager := credentials.NewManager()
+			mockCredsMetadataSetter := mock_session.NewMockCredentialsMetadataSetter(ctrl)
+			switch tc.roleType {
+			case credentials.ApplicationRoleType:
+				iamRoleCredentialsMessageCopy.RoleType = aws.String(credentials.ApplicationRoleType)
+				mockCredsMetadataSetter.EXPECT().
+					SetTaskRoleMetadata(gomock.Any()).
+					Return(nil)
+			case credentials.ExecutionRoleType:
+				iamRoleCredentialsMessageCopy.RoleType = aws.String(credentials.ExecutionRoleType)
+				mockCredsMetadataSetter.EXPECT().
+					SetExecRoleMetadata(gomock.Any()).
+					Return(nil)
+			default:
+				t.Fatal("invalid role type used in happy path test, role type should be valid for happy path")
+				return
+			}
+			mockMetricsFactory := mock_metrics.NewMockEntryFactory(ctrl)
+			mockEntry := mock_metrics.NewMockEntry(ctrl)
+			mockEntry.EXPECT().WithFields(gomock.Any()).Return(mockEntry)
+			mockEntry.EXPECT().Done(nil)
+			mockMetricsFactory.EXPECT().New(metrics.CredentialsRefreshSuccess).
+				Do(func(arg0 interface{}) {
+					defer wg.Done() // decrement WaitGroup counter now that HandleResourceAttachment function has been called
+				}).
+				Return(mockEntry)
+
+			testResponseSender := func(response interface{}) error {
+				resp := response.(*ecsacs.IAMRoleCredentialsAckRequest)
+				ackSent <- resp
+				return nil
+			}
+			testRefreshCredentialsResponder := NewRefreshCredentialsResponder(credentialsManager,
+				mockCredsMetadataSetter,
+				mockMetricsFactory,
+				testResponseSender)
+
+			handleCredentialsMessage :=
+				testRefreshCredentialsResponder.HandlerFunc().(func(*ecsacs.IAMRoleCredentialsMessage))
+
+			go handleCredentialsMessage(&iamRoleCredentialsMessageCopy)
```

Comment:
Why do we want to call `handleCredentialsMessage` asynchronously?

Reply:
I actually don't think calling asynchronously is necessary or adds any benefit here in this case - I'll remove this. Thanks for calling out!

Reply:
Thanks